### PR TITLE
feat: add bullet list element parsing and rendering

### LINKED ISSUE #92: [REQ] Impl bullet list parsing and rendering
List parsing will be another major step, since it is the first element where indentation is important.
The `bullet list` is the simplest list, so it is best suited to start with it.

### PR DESCRIPTION
**Pull request by mhatzl**
## List of issues that this PR closes

closes #92 

### Removed `Blankline` symbol

The `Blankline` symbol was used to indicate two consecutive newlines.
This created problems with prefix matching, because prefix is only matched after newlines.
As a better alternative, the `SymbolIterator` offers `is_empty_line()` and `consumed_is_empty_line()`.
Those functions also handle lines only consisting of whitespace symbols.

### Adapt `peeking_next()` to consider matching functions

With this change, `peeking_next()` no returns the same symbols as `next()`.
This is important for nested elements, because matching functions internally use `peeking_next()`.
Meaning that matching functions of nested elements would not be correct if `peeking_next()` would not consider the matching functions of outer iterators.

### CODE REVIEW

File: commons/src/scanner/symbol/iterator/mod.rs [hunk 1]
```diff
@@ -304,14 +349,54 @@ impl<'input> PeekingNext for SymbolIterator<'input> {
...
+
+        if !self.next_matching && !self.peek_matching {
+            if let Some(end_fn) = self.end_match.clone() {
+                let peek_index = self.peek_index();
+                self.peek_matching = true;
 
-        match &mut self.kind {
+                let end_matched = (end_fn)(self);
+
+                self.peek_matching = false;
```
Comment (nfejzic) [inline commons/src/scanner/symbol/iterator/mod.rs:363]: Ok, I think I get how these `*_matching` fields are used. Basically, that's the context for the matcher function to know whether we're in `peek` or `next` function call? 

If that's the case, it seems like it would be to forget changing these variables (always the case with mutating state). We could maybe use a different solution:

1. Change the signature of matcher function to take additional `MatcherContext` parameter: 
   * `MatcherContext` has functions or fields for checking which `*_matching` it is etc. 
   * It should be easy to create, maybe builder pattern
   * Benefit is, we cannot forget to create it, since function needs it as parameter
2. Do not implement `*Matcher` trait for `SymbolIterator` directly, rather wrap iterator in another struct that implements that trait 
   * this wrapper struct can have fields indicating what the context is
   * we implement `*Mathcer` trait for it
   * Benefit is, `SymbolIterator` cannot be passed into the matcher functions without being wrapped in matcher struct 
   * Might be a little harder to implement, maybe with generics? 
   * Might also be a little less flexible than using a second parameter for context, but maybe that's a good thing

Both of these options would probably require a little more boilerplate (another struct, implementation of the struct etc.), but I think it would improve code readability and also prevent some potential mistakes.

File: commons/src/scanner/symbol/iterator/mod.rs [hunk 2]
```diff
@@ -304,14 +349,54 @@ impl<'input> PeekingNext for SymbolIterator<'input> {
...
+                self.set_peek_index(peek_index);
+
+                if end_matched {
+                    return None;
+                }
+            }
+        }
+
+        let peeked_symbol_opt = match &mut self.kind {
```
Comment (nfejzic) [inline commons/src/scanner/symbol/iterator/mod.rs:372]: I'm also questioning our decision to call this `kind`. Maybe `base`, `parent` or `wrapped`? In either case, there is some underlying iterator inside, not just kind 🤔.

But I don't think the name choice is too important here.

File: parser/src/elements/atomic/paragraph.rs [hunk 3]
```diff
@@ -54,21 +55,24 @@ impl From<Vec<&'_ Symbol<'_>>> for Paragraph {
...
-
-        let output = TokenizeOutput { tokens: content };
+        let mut content_iter = input.nest(
+            None,
+            Some(Rc::new(|matcher: &mut dyn EndMatcher| {
+                matcher.consumed_is_empty_line()
+            })),
+        );
+        let content = content_iter.take_to_end();
+        content_iter.update(input);
```
Comment (nfejzic) [inline parser/src/elements/atomic/paragraph.rs:69]: It just occurred to me how could we (maybe) enforce update of `input` iterator. 

We could make it so that `take_to_end` function takes the `input` iterator as parameter. The function returns `Option<Vec<Symbol>>`: 

1. Consumes symbols until the end
2. Checks whether the end is reached
   * if end is reached
     - update `input` iterator
     - `return Some(symbols)`
   * if end is not reached 
     - do not update `input` iterator
     - `return None`

Then, to get the content, we have to provide the original input iterator. And we can then automatically update the input iterator, preventing the bug where we might forget to update the input iterator. What do you think? 🤔

File: parser/src/elements/indents/bullet_list.rs [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+impl ElementParser for BulletListEntry {
+    type Token<'a> = self::EntryToken;
+
+    fn tokenize<'i>(
+        input: &mut unimarkup_commons::scanner::SymbolIterator<'i>,
+    ) -> Option<crate::TokenizeOutput<Self::Token<'i>>> {
+        let entry_keyword = BulletListEntryKeyword::try_from(input.next()?).ok()?;
+
+        if SymbolKind::Whitespace != input.next()?.kind {
```
Comment (nfejzic) [inline parser/src/elements/indents/bullet_list.rs:186]: I would reverse this, find it easier to read: 

```suggestion
        if input.next()?.kind != SymbolKind::Whitespace {
```

**Review by nfejzic — COMMENTED** [review]
Awesome!

I think we now have a fairly decent parsing framework. Looks like it's pretty flexible and not that hard to work with, and that makes me really happy. 

As for the PR itself, I left some comments. Many of them are not directly related to this PR. You can decide for yourself, whether we should tackle this in this PR. 

I am almost leaning towards merging this and then tackling the comments in other PRs so that we prevent explosion in number of changes (is already 1+k LOC large, would not like to inflate it even more). So, I will just leave comments for now (neither approved nor changes requested) so we can discuss first how should we go forward.

File: commons/src/scanner/symbol/iterator/mod.rs [hunk 5]
```diff
@@ -34,6 +34,13 @@ pub struct SymbolIterator<'input> {
     end_match: Option<IteratorEndFn>,
     /// Flag set to `true` if this iterator reached its end.
     iter_end: bool,
+    /// Flag set to `true` if prefix mismatch occured.
+    prefix_mismatch: bool,
+    /// Flag set to `true` to indicate matching context in [`Self::next()`].
+    next_matching: bool,
+    /// Flag set to `true` to indicate matching context in [`Self::peeking_next()`]
+    /// Used to prevent consumed matching while peeking.
+    peek_matching: bool,
```
[the remark this answers is not in the document]
Comment (mhatzl) [reply commons/src/scanner/symbol/iterator/mod.rs:43]: I decided to have this info stored as state, because I don't see an easy way to handle this per function. The state option makes it easier with nesting iterators, because it does not affect parent iterators.
Because this state handling must only be taken into account in the iterator and matcher implementation, but not in any usage of the iterator, I think it is acceptable to have this complexity here. I added tests to ensure the behavior of peek() and next() are the same, and I don't assume that the implementation will change much in the future.

File: commons/src/scanner/symbol/iterator/mod.rs [hunk 6]
```diff
@@ -262,13 +300,16 @@ impl<'input> Iterator for SymbolIterator<'input> {
...
+        if self.prefix_mismatch || self.end_reached() {
             return None;
         }
 
         if let Some(end_fn) = self.end_match.clone() {
+            self.next_matching = true;
+
             if (end_fn)(self) {
                 self.iter_end = true;
+                self.next_matching = false;
```
[the remark this answers is not in the document]
Comment (mhatzl) [reply commons/src/scanner/symbol/iterator/mod.rs:312]: I updated the doc on `next_matching`. Hopefully this makes this code clearer.

In short: Setting `next_matching` flag in `next()` prevents matching in `peeking_next()` during execution of the matching functions in `next()`. Matching functions must get the symbols as is, because they define what symbols must be skipped.

File: commons/src/scanner/symbol/iterator/mod.rs (hunk 1 above)
Comment (mhatzl) [reply commons/src/scanner/symbol/iterator/mod.rs:363]: I want to keep the `Matcher` traits as is, because these are used in all elements.
Adding complexity here would spread complexity to all element implementations.
With the internal flags, the complexity is bound to the two files in the `SymbolIterator` implementation. I think it is better to have this complexity here with the chance of getting it wrong, because it makes it easier on the element implementation side.

File: commons/src/scanner/symbol/iterator/mod.rs (hunk 2 above)
Comment (mhatzl) [reply commons/src/scanner/symbol/iterator/mod.rs:372]: `parent` sounds good

File: parser/src/elements/atomic/paragraph.rs (hunk 3 above)
Comment (mhatzl) [reply parser/src/elements/atomic/paragraph.rs:69]: It must be able to call `take_to_end()` without passing an iterator, because it is possible to set end matching without nesting an iterator.

I would leave it as is right now.
Maybe once we implement more elements we see a better option on how to enforce the update.

File: commons/src/scanner/symbol/iterator/mod.rs (hunk 1 above)
Comment (nfejzic) [reply commons/src/scanner/symbol/iterator/mod.rs]: Well, we could have both of those benefits by wrapping the iterator in a separate type that implements `Matcher` trait. This separate type could have a reference to the iterator internally, and additional needed context. We create it when we call the function, and it is dropped right after. That way you can't make a mistake by forgetting to reset state. 

It would be something like this: 

```rust
// instead of this 
self.peek_matching = true;
let end_matched = (end_fn)(self);
self.peek_matching = false;

// we would do something like
let end_matched = (end_fn)(Matcher { iter: self, state: MatcherState::Peek });

// where Matcher is something like: 
struct Matcher<'iter> {
    iter: &'iter mut SymbolIterator<'iter>, // or whatever necessary
    state: MatcherState, 
}

enum MatcherState {
    Peek,
    Next,
}
```

`Matcher` then just implements the `*Matcher` traits. The complexity would still be contained in iterator implementation and users of iterator and matching functions do not know about this at all, but the benefit is we do not have to keep track of state in iterator.

File: commons/src/scanner/symbol/iterator/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (nfejzic) [reply commons/src/scanner/symbol/iterator/mod.rs]: I agree that it will probably stay like it is and won't change much. But at the same time, it is not easy to understand, and we probably will need to understand it again from time to time. So if we can improve that aspect of the code I would like to push it a little bit 😛. 

It might be the case that it's not really possible to improve that, since it's not that easy to solve anyway. But I think it's worth a try in any case 👀. 

TLDR: Complexity is (of course) acceptable, but if it's possible to reduce it then we should try anyway 👀.

[the remark this answers is not in the document]
Comment (mhatzl) [reply commons/src/scanner/symbol/iterator/mod.rs]: As internally discussed, we will move this optimization into its own issue.
Changes would be inside the symbol iterator, so merging this PR first helps to speed up development of other features.

File: commons/src/scanner/symbol/iterator/mod.rs (hunk 1 above)
Comment (mhatzl) [reply commons/src/scanner/symbol/iterator/mod.rs]: I will create an issue for this once I fix the issue templates.

**Review by nfejzic — APPROVED** [review]
